go/proxyd/daisychain.go

```diff
@@ -0,0 +1,416 @@
+package proxyd
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math/big"
+	"net/http"
+	"reflect"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/gorilla/mux"
+	"github.com/gorilla/websocket"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/rs/cors"
+)
+
+var (
+	// Represents the chain id for the Optimism L2 mainnet
+	MainnetChainId = big.NewInt(10)
+	// Represents the chain id for the Optimism L2 public testnet
+	KovanChainId = big.NewInt(69)
+)
+
+type DaisyChainServer struct {
+	rpcServer          *http.Server
+	maxBodySize        int64
+	authenticatedPaths map[string]string
+	epoch1             *Backend
+	epoch2             *Backend
+	epoch3             *Backend
+	epoch4             *Backend
+	epoch5             *Backend
+	epoch6             *Backend
+	bedrockCutoffBlock *big.Int
+	l2ChainID          *big.Int
+	upgrader           *websocket.Upgrader
+}
+
+type RequestOptions struct {
+	Epoch *uint `json:"epoch,omitempty"`
+}
+
+var latestEpoch = uint(6)
+
+func NewDaisyChainServer(
+	backends map[string]*Backend,
+	maxBodySize int64,
+	authenticatedPaths map[string]string,
+	l2ChainID *big.Int,
+	bedrockCutoffBlock *big.Int,
+) *DaisyChainServer {
+	srv := DaisyChainServer{
+		epoch1:             backends["epoch1"],
+		epoch2:             backends["epoch2"],
+		epoch3:             backends["epoch3"],
+		epoch4:             backends["epoch4"],
+		epoch5:             backends["epoch5"],
+		epoch6:             backends["epoch6"],
+		maxBodySize:        maxBodySize,
+		authenticatedPaths: authenticatedPaths,
+		bedrockCutoffBlock: bedrockCutoffBlock,
+		l2ChainID:          l2ChainID,
+		upgrader: &websocket.Upgrader{
+			HandshakeTimeout: 5 * time.Second,
+		},
+	}
+	return &srv
+}
+
+func StartDaisyChain(config *Config) (func(), error) {
+	if err := config.ValidateDaisyChainBackends(); err != nil {
+		return func() {}, err
+	}
+
+	lim := NewLocalRateLimiter()
+	_, backendsByName, err := config.BuildBackends(lim)
+	if err != nil {
+		return func() {}, err
+	}
+
+	resolvedAuth, err := config.ResolveAuth()
+	if err != nil {
+		return func() {}, err
+	}
+
+	// parse the config
+	srv := NewDaisyChainServer(
+		backendsByName,
+		config.Server.MaxBodySizeBytes,
+		resolvedAuth,
+		config.Eth.L2ChainID,
+		config.Eth.BedrockCutoffBlock,
+	)
+
+	// send a chain id request to each node to ensure they are on the same chain
```

Comment:
This should be broken out into smaller functions so that all this business logic doesn't have to be in the same function